### ISSUE TITLE
Async MVP

### ISSUE THREAD

**Issue by paulkernfeld** [deffen3/discovery#6]
I started out by trying to copy and modify the lsm303dhlc crate, but the code is out of date and overly complex so I gave up. I'm going to just use the solution from chapter 14 instead.

I also realized we'll need to be careful not to let peripherals get into an async state. For example, if we're currently waiting for data from the magnetometer we definitely don't want to allow another task to send a new request for data to the magnetometer.

**Comment by paulkernfeld** [comment]
Okay, I am realizing that using Rust's built-in Future trait is going to be a bit tricky. This is because, when a Future is polled and it's not ready, the Future is responsible for waking up the task later. That is to say, if an executor sees that a Future isn't ready, it isn't required to keep polling that future. The challenge is that we don't currently have a good way to implement waking.

Looking at the implementation of the executor crate, I believe that [it does actually keep polling the main future forever](https://github.com/richardanaya/executor/blob/efebe660862d6a76a4a6a5ba3541a68aa00fd9f8/executor/src/lib.rs#L67-L77). This isn't to say that the executor crate is wrong; rather, the executor crate can (I think) successfully execute a future that doesn't do any waking. So, we might be able to get away with just not implementing any waking. However, I don't think we should go this route because we'd be implementing futures wrong.

One solution would be to use the Future defined in the WIP chapter of the book. That definition has no concept of waking and does need to be continuously polled, which will work well for our use case.

A weird but correct solution, suggested [here](https://users.rust-lang.org/t/polling-in-new-era-futures/30531/2?u=occupy_paul_st), would be to have our implementations of future always immediately call wake when poll returns NotReady. I might try to do this.

If we can get interrupts working, that should let us implement waking. This would be cool but is probably harder.

**Comment by paulkernfeld** [comment]
I'm still kind of interested in the immediately-wake solution, but the executor crate has a bug that prevents it from working (https://github.com/richardanaya/executor/issues/4). So, we'd need to find an executor that does or implement our own executor.

**Comment by mjhoy** [comment]
dang, this all seems really interesting. i haven't really dug in yet. are you streaming this tomorrow?

**Comment by paulkernfeld** [comment]
Yeah, I think so! I might try to write a custom terrible block_on.

**Comment by paulkernfeld** [comment]
I have been enjoying this a lot, and definitely learning a lot about Rust's async ecosystem. I implemented some Future stuff on my Twitch stream this morning, and this afternoon I actually figured out how to get streams working! I put the code into a separate repo, [here](https://github.com/paulkernfeld/compass/blob/master/src/main.rs). If you guys want to use Rust's built-in async/await, you could copy/paste the `inefficient` module that I wrote.

My next step might be to try to add the timer as a separate stream. Then, I could use something like `stream::select` from the `futures` crate to make a reactive app that executes some logic each time it receives an input event from the compass or timer.
